File: mysql-async/src/test/scala/com/github/mauricio/async/db/mysql/TransactionSpec.scala
package com.github.mauricio.async.db.mysql

import java.util.UUID
import java.util.concurrent.TimeUnit

import org.specs2.mutable.Specification
import com.github.mauricio.async.db.util.FutureUtils.awaitFuture
import com.github.mauricio.async.db.mysql.exceptions.MySQLException
import com.github.mauricio.async.db.Connection

import scala.concurrent.duration.Duration
import scala.concurrent.{Await, Future}
import scala.util.{Success, Failure}

object TransactionSpec {

  val BrokenInsert = """INSERT INTO users (id, name) VALUES (1, 'Maurício Aragão')"""
  val InsertUser = """INSERT INTO users (name) VALUES (?)"""
  val TransactionInsert = "insert into transaction_test (id) values (?)"

}

class TransactionSpec extends Specification with ConnectionHelper {

  import TransactionSpec._

  "connection in transaction" should {

    "correctly store the values of the transaction" in {
      withConnection {
        connection =>
          executeQuery(connection, this.createTable)

          val future = connection.inTransaction {
            c =>
            c.sendPreparedStatement(this.insert)
             .flatMap( r => connection.sendPreparedStatement(this.insert))
          }

          awaitFuture(future)

          val result = executePreparedStatement(connection, this.select).rows.get
          result.size === 2

          result(0)("name") === "Maurício Aragão"
          result(1)("name") === "Maurício Aragão"
      }
    }

    "correctly rollback changes if the transaction raises an exception" in {

      withConnection {
        connection =>
          executeQuery(connection, this.createTable)
          executeQuery(connection, this.insert)

          val future = connection.inTransaction {
            c =>
              c.sendQuery(this.insert).flatMap(r => c.sendQuery(BrokenInsert))
          }

          try {
            awaitFuture(future)
            failure("should not have arrived here")
          } catch {
            case e : MySQLException => {

              e.errorMessage.errorCode === 1062
              e.errorMessage.errorMessage === "Duplicate entry '1' for key 'PRIMARY'"

              val result = executePreparedStatement(connection, this.select).rows.get
              result.size === 1
              result(0)("name") === "Maurício Aragão"
              success("correct result")
            }
          }
      }

    }

    "should make a connection invalid and not return it to the pool if it raises an exception" in {

      withPool {
        pool =>

          executeQuery(pool, this.createTable)
          executeQuery(pool, this.insert)

          var connection : Connection = null

          val future = pool.inTransaction {
            c =>
              connection = c
              c.sendQuery(BrokenInsert)
          }

          try {
            awaitFuture(future)
            failure("this should not be reached")
          } catch {
            case e : MySQLException => {

              pool.availables must have size(0)
              pool.availables must not contain(connection.asInstanceOf[MySQLConnection])

              success("success")
            }
          }

      }

    }

    "runs commands for a transaction in a single connection" in {

      val id = UUID.randomUUID().toString

      withPool {
        pool =>
          val operations = pool.inTransaction {
            connection =>
              connection.sendPreparedStatement(TransactionInsert, List(id)).flatMap {
                result =>
                  connection.sendPreparedStatement(TransactionInsert, List(id)).map {
                    failure =>
                      List(result, failure)
                  }
              }
          }

          Await.ready(operations, Duration(5, TimeUnit.SECONDS))

          operations.value.get match {
            case Success(e) => failure("should not have executed")
            case Failure(e) => {
              e.asInstanceOf[MySQLException].errorMessage.errorCode === 1062
              executePreparedStatement(pool, "select * from transaction_test where id = ?", id).rows.get.size === 0
              success("ok")
            }
          }

      }

    }

  }

}

File: mysql-async/src/test/java/com/github/jasync/sql/db/mysql/TransactionSpec.kt
package com.github.jasync.sql.db.mysql

import com.github.jasync.sql.db.invoke
import com.github.jasync.sql.db.mysql.exceptions.MySQLException
import com.github.jasync.sql.db.util.ExecutorServiceUtils
import com.github.jasync.sql.db.util.flatMapAsync
import com.github.jasync.sql.db.util.mapAsync
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.util.UUID
import java.util.concurrent.ExecutionException

val BrokenInsert = """INSERT INTO users (id, name) VALUES (1, 'Maurício Aragão')"""
val TransactionInsert = "insert into transaction_test (id) values (?)"

class TransactionSpec : ConnectionHelper() {

    @Test
    fun `connection in transaction should correctly store the values of the transaction`() {
        withConnection { connection ->
            executeQuery(connection, this.createTable)

            val future = connection.inTransaction { c ->
                c.sendPreparedStatement(this.insert)
                    .flatMapAsync(ExecutorServiceUtils.CommonPool) { connection.sendPreparedStatement(this.insert) }
            }

            future.get()

            val result = executePreparedStatement(connection, this.select).rows
            assertThat(result.size).isEqualTo(2)

            assertThat(result[0]["name"]).isEqualTo("Boogie Man")
            assertThat(result[1]("name")).isEqualTo("Boogie Man")
        }
    }

    private fun verifyException(
        exType: Class<out java.lang.Exception>,
        causeType: Class<out java.lang.Exception>? = null,
        body: () -> Unit
    ): Throwable {
        try {
            body()
            throw Exception("${exType.simpleName}->${causeType?.simpleName} was not thrown")
        } catch (e: Exception) {
            // e.printStackTrace()
            assertThat(e::class.java).isEqualTo(exType)
            causeType?.let { assertThat(e.cause!!::class.java).isEqualTo(it) }
            return e.cause ?: e
        }
    }

    @Test
    fun `connection in transaction should correctly rollback changes if the transaction raises an exception`() {
        withConnection { connection ->
            executeQuery(connection, this.createTable)
            executeQuery(connection, this.insert)

            val future = connection.inTransaction { c ->
                c.sendQuery(this.insert)
                    .flatMapAsync(ExecutorServiceUtils.CommonPool) { c.sendQuery(BrokenInsert) }
            }

            val e: MySQLException = verifyException(ExecutionException::class.java, MySQLException::class.java) {
                awaitFuture(future)
            } as MySQLException

            assertThat(e.errorMessage.errorCode).isEqualTo(1062)
            assertThat(e.errorMessage.errorMessage).matches("Duplicate entry '1' for key '(users\\.)?PRIMARY'")

            val result = executePreparedStatement(connection, this.select).rows
            assertThat(result.size).isEqualTo(1)
            assertThat(result[0]("name")).isEqualTo("Boogie Man")
        }
    }

    @Test
    fun `connection in transaction should should make a connection invalid and not return it to the pool if it raises an exception`() {
        withPool { pool ->

            executeQuery(pool, this.createTable)
            executeQuery(pool, this.insert)

            val future = pool.inTransaction { c ->
                c.sendQuery(BrokenInsert)
            }

            verifyException(ExecutionException::class.java, MySQLException::class.java) {
                awaitFuture(future)
            } as MySQLException

            assertThat(pool.idleConnectionsCount).isEqualTo(0)
            // pool.availables must not contain (connection.asInstanceOf[MySQLConnection])
        }
    }

    @Test
    fun `connection in transaction should runs commands for a transaction in a single connection`() {
        val id = UUID.randomUUID().toString()

        withPool { pool ->
            val operations = pool.inTransaction { connection ->
                connection.sendPreparedStatement(TransactionInsert, listOf(id))
                    .flatMapAsync((ExecutorServiceUtils.CommonPool)) { result ->
                        connection.sendPreparedStatement(TransactionInsert, listOf(id))
                            .mapAsync(ExecutorServiceUtils.CommonPool) { failure ->
                                listOf(result, failure)
                            }
                    }
            }

            val e: MySQLException = verifyException(ExecutionException::class.java, MySQLException::class.java) {
                operations.get()
            } as MySQLException

            assertThat(e.errorMessage.errorCode).isEqualTo(1062)
            val rows = executePreparedStatement(pool, "select * from transaction_test where id = ?", listOf(id)).rows
            assertThat(rows).isEmpty()
        }
    }

    @Test
    fun `check auto-commit and in transaction flag`() {
        withConnection { connection ->
            assertThat(connection.isAutoCommit()).isTrue()
            awaitFuture(connection.sendQuery("SET AUTOCOMMIT=0"))
            assertThat(connection.isAutoCommit()).isFalse()
        }
    }
}
